src/components/player/now-playing/details/switcher.ts

```diff
@@ -0,0 +1,270 @@
+  }
+
+  private get_action_group() {
+    const action_group = new Gio.SimpleActionGroup();
+
+    (action_group.add_action_entries as AddActionEntries)([
+      {
+        name: "select-page",
+        parameter_type: "s",
+        state: `"${this.get_details_id()}"`,
+        activate: (source, parameter) => {
+          const name = parameter?.get_string()[0];
+          if (name === undefined) return;
+
+          this.show_details = this.details_stack.visible_child_name === name
+            ? !this.show_details
+            : true;
+          this.details_stack.visible_child_name = name;
+
+          // chore: updating state
+          source.set_state(
+            GLib.Variant.new_string(this.get_details_id()),
+          );
+        },
+      },
+    ]);
+
+    return action_group;
+  }
+
+  private update_page_state() {
+    this.action_group.action_state_changed(
+      "select-page",
+      GLib.Variant.new_string(this.get_details_id()),
+    );
+  }
+
+  private get_details_id() {
+    if (!this.show_details || !this.details_stack) return "none";
+    return this.details_stack.visible_child_name;
+  }
+
+  private pages_listener = new SignalListeners();
+  private switcher_listener = new SignalListeners();
+
+  get details_stack() {
+    return this._details_stack;
+  }
+
+  set details_stack(value: Adw.ViewStack) {
+    if (value === this._details_stack) return;
+
+    this._details_stack = value;
+    this.init_switchers();
+  }
+
+  get pages() {
+    return this._details_stack.pages as Adw.ViewStackPages<Adw.ViewStackPage>;
+  }
+
+  private recreate_switchers() {
+    this.switcher_listener.clear();
+
+    // remove all children
+    let child: Gtk.Widget | null;
+    while (child = this.get_first_child()) child.unparent();
+
+    // create already existing pages
+    list_model_to_array(this.pages).forEach((page) => {
+      const button = this.create_switcher_button(page);
+      this.append(button);
+      return button;
+    });
+  }
+
+  private init_switchers() {
+    if (!this._details_stack) return;
+    this.recreate_switchers();
+
+    this.pages_listener.clear();
+
+    this.pages_listener.connect(
+      this.pages,
+      "items-changed",
+      this.pages_items_changed_cb.bind(this),
+    );
+
+    this.pages_listener.connect(this.pages, "selection-changed", () => {
+      this.update_page_state();
+    });
+
+    this.pages_listener.connect(this, "notify::show-details", () => {
+      this.update_page_state();
+    });
+
+    this.update_page_state();
+  }
+
+  private buttons = new Gio.ListStore<Gtk.Button>({
+    item_type: Gtk.Button.$gtype,
+  });
+
+  private pages_items_changed_cb(
+    _model: Gio.ListModel,
+    position: number,
+    removed: number,
+    added: number,
+  ) {
+    // remove buttons for removed pages
+    const removed_buttons = list_model_to_array(this.buttons).splice(
+      position,
+      position + removed,
+      ...(
+        new Array(added)
+          .map((index) => {
```

Comment:
```suggestion
        new Array(added)
          .fill()
          .map((_, index) => {
```

The array constructor initializes an array of empty values, which are ignored when trying to map

![image](https://github.com/vixalien/muzika/assets/8540562/707fb831-2152-4f86-97df-692c0dc08599)